Parallel execution docs

grasshopper7 opened the pull request:
Added for [Document how to run tests in parallel](https://github.com/cucumber/docs.cucumber.io/issues/298) issue. Added two new files - parallel-execution.md and parallel-timeline-report.png.

For java only but I think with some changes could be adapted for kotlin. Have used Maven only for dependencies and execution.  For Grade need to figure it out.

Maybe need a repository to store the source code files. Though it is pretty bare bones basic stuff.

grass-hopper-moc committed:
Added Timeline report image

grass-hopper-moc committed:
Added Parallel Intro

grass-hopper-moc committed:
Added JUnit part

grass-hopper-moc committed:
Added TestNG part

grass-hopper-moc committed:
Added CLI part

mlvandijk commented:
Thanks @grasshopper7 - I'll have to have a look when I have a little more time, so I can try this out.

mpkorstanje commented:
Quick remarks, sorry for being terse, on mobile:

1. Don't use `M` versions. They're release mile stones and still unstable.
2. Listing all versions of all jars is unmaintainable and prone to change.
3. Don't use a `feature` folder. Use java conventions and put the features in the same package as the runner. E.g. `com.example.app.RunCucumberTest` will automatically find all features in `src/test/resources/com/example/app`. This also removes some config.

mpkorstanje commented:
4. Same for the glue. If its in a descendant package from the runner glue need not be configured. Using these conventions helps keep the docs simple.

mpkorstanje commented:
5. Use Maven conventions and make sure the runner class ends with `Test.java`. E.g. `RunCucumberTest.java`. Maven will pick this up automatically. So then the config that references `Runner.java` can also be removed.

mpkorstanje commented:
6. I would leave out the failsafe config all together. Just link the relevant maven documentation. This is Cucumber docs, not maven docs.

mpkorstanje commented:
7. Looks pretty good!

aslakhellesoy commented:
Can you use hugo tags for versions? See the general cucumber-jvm install doc for examples

grass-hopper-moc committed:
Changes done

grass-hopper-moc committed:
Updated community link

grasshopper7 commented:
@mpkorstanje  Have updated the changes except for point 2.

Point 2 - _Listing all versions of all jars is unmaintainable and prone to change._ - Yep knew this one is gonna be a mess. Thought about mentioning latest version but the issue is with 4.3.0, it gets gherkin 5.1 and cucumber-expression 6.6.2, which are are not latest versions in repo.  Maybe mention something like - "Use the IDE to figure out transitive dependency versions of cucumber-java8 to add in classpath". Any thoughts?

Point 6 - _I would leave out the failsafe config all together._ - Think it makes things clearer regarding failsafe configurations. Have kept the junit config and referred it from the testng part. Added links to community page of tools containing maven execution details inside the docs.

Point 3,4,5 - Done simplifies things a lot.

Point 1 - _Don't use M versions. They're release mile stones and still unstable_. - Updated.

grasshopper7 commented:
@aslakhellesoy 
> Can you use hugo tags for versions? See the general cucumber-jvm install doc for examples

I added the version tags but it is showing **4.2.6** instead of latest **4.3.0**. Any pointers how to fix this. 

Even in installation page it is showing 4.2.6 but 4.3.0 for snapshot.

grass-hopper-moc committed:
Updated cli command formatting

mlvandijk commented:
@grasshopper7 The versions are specified in [data/versions.yaml](https://github.com/cucumber/docs.cucumber.io/blob/master/data/versions.yaml). Whenever there is a new release, they *should* be updated, but sometimes we forget (as docs and code are separate projects). I' ve just updated the versions in [this commit](https://github.com/cucumber/docs.cucumber.io/commit/43f28aec1b8d8ac0dbc79981138a66042790a745), so it should be ok now :)

grass-hopper-moc committed:
Added kotlin portion

grasshopper7 commented:
@mlvandijk Version numbers are now updated.
I added the kotlin part. First time I tried anything in kotlin. Maybe there is a better approach.  Thanks.

bischoffdev commented:
This is a great start!
However, there are some points I want to mention (quickly read this document, I hope I will have more time later to do a more in-depth review) :-) 

* for the Maven dependencies it might be better to not specify concrete version numbers (in the document you use 4.2.6 but 4.3.0 is already available)
* the thread IDs are not fixed (not always 13 and 14 like in the example) so this could be misleading
* there is actually another way to run the tests by having the configuration directly inside of a JUnit runner like so:
```
import cucumber.api.cli.Main;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class RestTest_IT {

    @Test
    public void test() {
        String[] arguments = new String[]{
                "--threads", "4",
                "-p", "json:target/json/test.json",
                "-p", "timeline:target/cucumber-parallel-report",
                "-p", "usage:target/cucumber-usage.json",
                "-p", "pretty",
                "-g", "your.glue.package",
                "path/to/features/"
        };
        int exitStatus = Main.run(arguments,
                Thread.currentThread().getContextClassLoader());
        assertThat(exitStatus, is(0));
    }
}
```

mlvandijk commented:
@laxersaz FYI: your first comment regarding Maven versions should be fixed by using the hugo tags (as mentioned above). @grasshopper7 - you might want to merge master into your branch ;)

grasshopper7 committed:
Merge pull request #1 from cucumber/master

Add changes

grass-hopper-moc committed:
Versions updated

grasshopper7 commented:
@laxersaz Added my comments inline to your observations. Thanks.

> This is a great start!
> However, there are some points I want to mention (quickly read this document, I hope I will have more time later to do a more in-depth review) :-)
> 
> * for the Maven dependencies it might be better to not specify concrete version numbers (in the document you use 4.2.6 but 4.3.0 is already available) -- **Fixed it.**

> * the thread IDs are not fixed (not always 13 and 14 like in the example) so this could be misleading -- **Think it is a Java thing regarding thread ids. Also it is mentioned, that the results would be similar rather than exact** 

> * there is actually another way to run the tests by having the configuration directly inside of a JUnit runner like so: -- **I think it belongs to a different guide dealing with cucumber-jvm execution methods. The run() method is called by the main() method (described in the CLI section), only the exit status is swallowed in the main() method. This can be used in a testng test method or in a java class also.**

grasshopper7 commented:
> @laxersaz FYI: your first comment regarding Maven versions should be fixed by using the hugo tags (as mentioned above). @grasshopper7 - you might want to merge master into your branch ;)

@mlvandijk  My bad. Fixed it. Though the version needs to be updated again. Shows 4.3.0, though 4.3.1 is released.

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+weight: 1800
+---
+
+Cucumber-JVM allows [parallel execution](https://cucumber.io/blog/announcing-cucumber-jvm-4-0-0/) across multiple threads since version 4.0.0. There are several options to incorporate this inbuilt feature in a Cucumber project.
```
```suggestion
Cucumber-JVM allows [parallel execution](https://cucumber.io/blog/announcing-cucumber-jvm-4-0-0/) across multiple threads since version 4.0.0. There are several options to incorporate this built-in feature in a Cucumber project.
```

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+- [TestNG](/docs/guides/parallel-execution/#testng)
+- [CLI](/docs/guides/parallel-execution/#cli)
+
+For each of these options, this tutorial will look at the project setup, configuration settings, execution commands.
```
```suggestion
For each of these options, this tutorial will look at the project setup, configuration settings and execution commands.
```

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+
+
+# JUnit
+Cucumber can be executed in parallel using **JUnit and Maven test execution plugins**. In JUnit the **feature files are run in parallel rather than scenarios**, which means **all the scenarios in a feature file will be executed by the same thread**. One can use either Maven Surefire or Failsafe plugin for executing the runners.
```
```suggestion
Cucumber can be executed in parallel using **JUnit and Maven test execution plugins**. In JUnit the **feature files are run in parallel rather than scenarios**, which means **all the scenarios in a feature file will be executed by the same thread**. You can use either Maven Surefire or Failsafe plugin to execute the runners.
```

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+
+	@BeforeStep
+	public void beforeStep() throws InterruptedException {
+		Thread.sleep(1000);
```
Why do we need this `sleep()` ?

grasshopper7 replied:
Ooops... I had added this to slow things down. Forgot to remove it. No need at all. Will remove it.

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,396 @@
+</dependency>
+```
+
+- Create a **parallel** folder in `src/test/resources` path and add the two feature files (`scenarios.feature` and `scenariooutlines.feature`) inside it.
````
Is the "parallel" folder needed (both in `src/test/resources` and `src/test/java`) for configuration reasons, or is this a nice to have / suggestion?

grasshopper7 replied:
No configuration reason. Just a package name I thought would be appropriate. Will mention that folder can be named anything.

mlvandijk replied:
Ok, thanks. That's what I thought, I just wanted to make sure. If it's not needed, maybe don't mention creating a folder at all, to prevent confusion?

grasshopper7 replied:
The class files are kept in the parallel package. If the feature files are kept in the parallel folder these are picked up automatically without adding any configuration to the CucumberOptions annotation.

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+		}
+		
+		Given("Step from {string} in {string} feature file") { scenario: String , file: String ->
+            println("Thread ID - %2d - %s from %s feature file.".format(Thread.currentThread().id,scenario,file))
```
Optional: Can we use String templating here? Maybe not with formatting... See also: https://stackoverflow.com/questions/23086291/format-in-kotlin-string-templates (it's fine to keep this as is)

grasshopper7 replied:
Done will add to commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,396 @@
+```
+{{% /block %}}
+
+- Add a cucumber **runner** using the `RunWith` annotation in the `parallel` package in the `src/test/java` folder.
````
same question here regarding configuration. Is `parallel` needed? (I.e. will it work if I don't put them in a directory / package named `parallel`)

grasshopper7 replied:
No configuration reason. Just a package name I thought would be appropriate. Will mention that folder can be named anything.

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,396 @@
+</plugin>
+```
+
+- Use the Maven `install` or a suitable command to **execute** the `POM`. This should run in parallel threaded mode. You should see a result similar to below. It is important to note that **both the scenarios in the file (`scenarios.feature`) are executed by thread with ID 14**. Similarly **both the rows of the scenario outline in the file (`scenariooutlines.feature`) are executed by thread with  ...
````
Rename to `scenario-outlines.feature` for readability? (if so, do this consistently)

grasshopper7 replied:
Will change all occurrences.

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,396 @@
+Thread ID - 14 - Scenario 2 from scenarios feature file.
+```
+
+- To execute using a Maven **Failsafe plugin include the below configuration** in the `build` section to the `POM`. Rename the runner class to `RunCucumberIT`.  You can find further details [here](/docs/community/not-cucumber/#maven-execution-plugin).
````
Ok, so if I understand correctly Surefire will run classes named *Test, and Failsafe will run classes named *IT?

grasshopper7 replied:
Yeah. The plugins pick up these class names matching the pattern automatically. There are a few other patterns mentioned in the linked article

mlvandijk replied:
Ok, thnx :)

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+
+If you have **multiple runners** then you can set the parallel option to `classesAndMethods`, `methods` or `classes`. For a single runner the `classes` option would be similar to a sequential execution.
+
+For a **visual representation** you can add the **timeline report** with the plugin option to a `CucumberOptions` annotation in the runner class. Scroll to the end for an image of the report.
```
I don't understand what you mean here? Could you be specific which option and how to add it?

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+	<groupId>org.apache.maven.plugins</groupId>
+	<artifactId>maven-failsafe-plugin</artifactId>
+	<version>2.22.0</version>
+	<executions>
```
I have added the `<executions>` section to our (Kotlin/Maven/JUnit) project and the build still passes. I'm not sure if and how I can see that the tests are run in parallel though?

mlvandijk replied:
I noticed that you are explicitly printing the Threads in this guide; ofcourse I don't do that in my project :)

grasshopper7 replied:
Thread id was the easiest I could think off. You can configure the timeline plugin to get a chart. But that would be bit off an overkill.

grasshopper7 replied:
Also the execution times should decrease.

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+For a **visual representation** you can add the **timeline report** with the plugin option to a `CucumberOptions` annotation in the runner class. Scroll to the end for an image of the report.
+
+
+# TestNG
```
There's some duplication in this section, compared to the previous section. I'm wondering if we could remove the duplication *without* making it confusing, but I'm not sure at this point...

grasshopper7 replied:
Maybe a separate section in this tutorial which mentions how to set this up and what is the result. Then can be linked from the other parts.

mlvandijk replied:
Much better now, thanks!

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+
+
+# CLI
+The `Main class` in the `cucumber.api.cli package` is used to execute the feature files. There is no need to create any runner class. The usage options for this class is mentioned [here](https://github.com/cucumber/cucumber-jvm/blob/v4.0.0/core/src/main/resources/cucumber/api/cli/USAGE.txt). The `--threads` option needs to be set to a value **greater than 1** to run in parallel. When the parallel ...
```
```suggestion
The `Main class` in the `cucumber.api.cli package` is used to execute the feature files. You can run this class directly from the command line; in that case, there is no need to create any runner class. The usage options for this class is mentioned [here](https://github.com/cucumber/cucumber-jvm/blob/v4.0.0/core/src/main/resources/cucumber/api/cli/USAGE.txt). The `--threads` option needs to be set to a value **greater than 1** to run in parallel. When the parallel mode is used, the scenarios and rows in a scenario outline will be run in multiple threads.
```

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,396 @@
+ 
+Below is the basic command to start the execution.
+```shell
+java -cp <classpath> cucumber.api.cli.Main -g <steps package> --threads <thread count> <path to feature files>
````
With `<steps package>`, do you mean the package containing the step definitions? In that case I'd prefer to use `<step defs package>` instead.

grasshopper7 replied:
How about the glue code package?

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,396 @@
+java -cp <classpath> cucumber.api.cli.Main -g <steps package> --threads <thread count> <path to feature files>
+```
+
+Follow the below steps to **execute the command from a terminal**.
````
```suggestion
Follow the steps below to **execute the command from a terminal**.
```

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+
+Follow the below steps to **execute the command from a terminal**.
+
+- Download the necessary cucumber jars from the cucumber.io Maven repository (links provided under version number) or, if available, copy from the local repository into a folder. These need to be provided in the **classpath**. The versions mentioned are compatible with the latest cucumber release 4.3.0.
```
Please, let's **not** describe downloading the jars, as we really recommend not doing that. I realise you are doing this here to give an alternative to getting the dependencies through Maven, but doing so leads to many problems and also I'd rather not maintain this list in the documentation.
In fact, in that case it might be better to also leave out the sections about the Maven dependencies for JUnit and TestNG, and just link to the part of the docs where that is already described. Which will save some duplication also :)

grasshopper7 replied:
Removed the dependencies in CLI section.
For the other two the 10 minute tutorial uses the cucumber-archetype. Is that updated to get latest versions? I tried it out around 3 months back and it added some pretty old version,

mlvandijk replied:
The archetype has been updated to a more recent version (v4.?) but probably not *the* most recent version... It's a great thing to use to get started, but a pain to maintain.

grasshopper7 replied:
Using the archetype is a good solution. As long as it gets a V4 it works, unless in a future version parallel functionality is removed. Will add the link.

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+	9.  `tag-expressions.jar` [`1.1.1`](http://central.maven.org/maven2/io/cucumber/tag-expressions/1.1.1/tag-expressions-1.1.1.jar)
+	10. `typetools.jar` [`0.5.0`](http://central.maven.org/maven2/net/jodah/typetools/0.5.0/typetools-0.5.0.jar)
+	
+- Create a folder named `cukecli` and a folder `parallel` inside it, to keep the step definition classes and feature files.
```
I'd prefer a `cucumber` or `cucumbercli` directory (personal preference, I guess)

grasshopper7 replied:
Will change it.

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+
+- Open up a **terminal window** and navigate to the source folder of the project, in our case `cukecli`.
+
+{{% block "java" %}}
```
When the content of the language blocks is the same, you can specify multiple languages in the same block, separated by a comma, like this `{{% text "java,kotlin" %}}` (see [Contributing to the documentation](https://cucumber.io/docs/community/contributing-to-documentation/#paragraphs-or-text-that-is-valid-for-multiple-programming-languages))

grasshopper7 replied:
Will change this

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,396 @@
+
+- Execute using the below command.
+
+{{% block "java" %}}
```
Same here.
Note that anything **outside** the language blocks would be visible on all polyglot pages (if we add javascript/ruby in the future).

grasshopper7 replied:
Will change this

mlvandijk reviewed (changes requested):
Hi @grasshopper7 - this is a more elaborate addition than I expected; thank you! Sorry it took my longer than I would've liked to look into it in more detail. Overall it's really great. I do have some minor textual changes, and am wondering if we could make sure to limit duplication (i.e. not say the same thing in multiple places, but add links).

grasshopper7 committed:
Update content/docs/guides/parallel-execution.md

Co-Authored-By: Marit van Dijk <mlvandijk@gmail.com>

grasshopper7 committed:
Update content/docs/guides/parallel-execution.md

Co-Authored-By: Marit van Dijk <mlvandijk@gmail.com>

grasshopper7 committed:
Update content/docs/guides/parallel-execution.md

Co-Authored-By: Marit van Dijk <mlvandijk@gmail.com>

grasshopper7 committed:
Update content/docs/guides/parallel-execution.md

Co-Authored-By: Marit van Dijk <mlvandijk@gmail.com>

grasshopper7 committed:
Update content/docs/guides/parallel-execution.md

Co-Authored-By: Marit van Dijk <mlvandijk@gmail.com>

grasshopper7 commented:
> Hi @grasshopper7 - this is a more elaborate addition than I expected; thank you! Sorry it took my longer than I would've liked to look into it in more detail. Overall it's really great. I do have some minor textual changes, and am wondering if we could make sure to limit duplication (i.e. not say the same thing in multiple places, but add links).

@mlvandijk I committed the minor text changes. Will have a look into the others later today.

luke-hill commented:
Just for continued / future reference @grasshopper7  (As this has gotten to be a heavily committed change), you're able to push up multiple commits in one go to the repo (Say 5 local commits in one go with a `git push`). Just so people monitoring don't need to check back as often

mlvandijk commented:
@luke-hill The multiple commits are probably due to @grasshopper7 accepting my suggested changes..

grasshopper7 commented:
> Just for continued / future reference @grasshopper7 (As this has gotten to be a heavily committed change), you're able to push up multiple commits in one go to the repo (Say 5 local commits in one go with a `git push`). Just so people monitoring don't need to check back as often

@luke-hill I accepted the changes suggested. Is there a way to avoid multiple commits for this? Thx.

Commit:
Changes from feedback

grasshopper7 commented:
@mlvandijk Made the changes other than the string templating and multiple runners.

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,389 @@
+
+- Create a Maven project in your favorite IDE. Add `cucumber-java8` and `cucumber-junit` **dependencies** to the `POM`.
+
+```shell
````
We also mention this [here](https://cucumber.io/docs/installation/java/#maven) and [here](https://cucumber.io/docs/tools/java/#maven) - I guess it's ok to duplicate though.

grasshopper7 replied:
Will add this and the archetype option. Reader can pick anyone of them.

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,389 @@
+	<version>{{% version "cucumberjvm" %}}</version>
+	<scope>test</scope>
+</dependency>
+<dependency>
```
We also mention this [here](https://cucumber.io/docs/cucumber/checking-assertions/#junit). Again, I think it's ok to duplicate, otherwise the reader would be clicking back and forth.

Note that the JUnit dependency itself is missing on this page - could you add it here too?

grasshopper7 replied:
No need to add the junit or testng dependency explicitly in pom. The cucumber-junit or cucumber-testng pulls them as a transitive one.

mlvandijk commented:
@grasshopper7 Could you merge master into this branch again (I think the versions have been updated again). I'll try to create a parallel project from this guide to make sure I get actual parallel execution and have a look to see if I can think of a way to prevent duplication on the page (to keep it easy to maintain in the future).

grasshopper7 committed:
Merge pull request #2 from cucumber/master

Update fork

grasshopper7 committed:
Merge branch 'master' into parallel-execution-docs

grasshopper7 committed:
Merge pull request #3 from grasshopper7/master

Update

grass-hopper-moc committed:
Review feedback changes

grasshopper7 commented:
@mlvandijk Removed the dependency addition for  JUnit and testng. Linked to existing documentatio. Updated kotlin string template. Removed duplication of timeline reporter to separate section. Tried to simplify multiple runners. 

Clicked the close button by mistake. Reopened.

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+weight: 1800
+---
+{{% block "java,kotlin" %}}
+Cucumber-JVM allows [parallel execution](https://cucumber.io/blog/announcing-cucumber-jvm-4-0-0/) across multiple threads since version 4.0.0. There are several options to incorporate this built-in feature in a Cucumber project.
```
```suggestion
Cucumber-JVM allows parallel execution across multiple threads since [version 4.0.0.](https://cucumber.io/blog/announcing-cucumber-jvm-4-0-0/) There are several options to incorporate this built-in feature in a Cucumber project. You can do so using:
```

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+{{% block "java,kotlin" %}}
+Cucumber can be executed in parallel using **JUnit and Maven test execution plugins**. In JUnit the **feature files are run in parallel rather than scenarios**, which means **all the scenarios in a feature file will be executed by the same thread**. You can use either Maven Surefire or Failsafe plugin to execute the runners.
+
+- Create a Maven project in your favorite IDE using the [cucumber-archetype](/docs/guides/10-minute-tutorial/#create-an-empty-cucumber-project) or adding Cucumber dependencies to the POM as detailed [here](https://cucumber.io/docs/installation/java/#maven) and Junit dependencies [here](/docs/cucumber/checking-assertions/#junit).
```
```suggestion
- Create a Maven project in your favorite IDE using the [cucumber-archetype](/docs/guides/10-minute-tutorial/#create-an-empty-cucumber-project) or by adding Cucumber dependencies to the POM as detailed [here](https://cucumber.io/docs/installation/java/#maven) and Junit dependencies [here](/docs/cucumber/checking-assertions/#junit).
```

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+
+	@Given("Step from {string} in {string} feature file")
+	public void step(String scenario, String file) {
+		System.out.format("Thread ID - %2d - %s from %s feature file.\n", Thread.currentThread().getId(), scenario,file);
```
```suggestion
		System.out.format("Thread ID - %2d - %s from %s feature file.\n",
		Thread.currentThread().getId(), scenario,file);
```

mlvandijk replied:
Prevent (or at least limit) horizontal scrolling.

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+# CLI
+
+{{% block "java,kotlin" %}}
+The `Main class` in the `cucumber.api.cli package` is used to execute the feature files. You can run this class directly from the command line; in that case, there is no need to create any runner class. The usage options for this class is mentioned [here](https://github.com/cucumber/cucumber-jvm/blob/v4.0.0/core/src/main/resources/cucumber/api/cli/USAGE.txt). The `--threads` option needs to be se ...
```
```suggestion
The `Main class` in the `cucumber.api.cli package` is used to execute the feature files. You can run this class directly from the command line; in that case, there is no need to create any runner class. The usage options for this class are mentioned [here](https://github.com/cucumber/cucumber-jvm/blob/v4.0.0/core/src/main/resources/cucumber/api/cli/USAGE.txt). The `--threads` option needs to be set to a value **greater than 1** to run in parallel. When the parallel mode is used, the scenarios and rows in a scenario outline will be run in multiple threads.
```

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,380 @@
+ 
+Below is the basic command to start the execution.
+```shell
+java -cp <classpath> cucumber.api.cli.Main -g <glue package> --threads <thread count> <path to feature files>
````
Could you please **also** provide an example of what this command would like look like with the setup as described in this guide (i.e. using directory `parallel`). I'm not too familiar with running Cucumber from the command line so I cannot get this to work.

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+
+Follow the steps below to **execute the command from a terminal**.
+
+- Download the cucumber jars from the cucumber.io repository into a folder. These need to be provided in the **classpath**.
```
This really needs to be removed. We don't recommend downloading individual jars, as it's very easy to get the wrong dependencies which leads to problems/questions.

mlvandijk replied:
In my opinion, we only need to mention how to **run** in parallel from the CLI, using the `--threads <number of threads>` option.

grasshopper7 replied:
Removed.

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+
+Follow the steps below to **execute the command from a terminal**.
+
+- Download the cucumber jars from the cucumber.io repository into a folder. These need to be provided in the **classpath**.
```
```suggestion
```

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+
+- Download the cucumber jars from the cucumber.io repository into a folder. These need to be provided in the **classpath**.
+	
+- Create a folder named `cucumbercli` and a folder `parallel` (or any folder names) inside it, to keep the step definition classes and feature files.
```
```suggestion
```

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+
+- Add the two feature files (`scenarios.feature` and `scenario-outlines.feature`) and **step definition class** as described in the JUnit section.
+
+- Open up a **terminal window** and navigate to the source folder of the project, in our case `cucumbercli`.
```
```suggestion
- Open up a **terminal window** and navigate to the source folder of the project.
```

grasshopper7 replied:
Added to batch commit

mlvandijk reviewed (commented):
Finally got around to check this guide step by step. Here is my progress so far:
[x] Java - Junit - Maven
[x] Java - TestNg - Maven
[ ] Java - Cli
I'm not too familiar with running from the CLI so I couldn't get that to work yet (see questions below).
If you could clarify, and/or someone else can confirm that it works, that would be great.

**Update:** also works with Kotlin, as long as StepDefs are in `src/test/**java**` (maybe it's possible to configure Surefire and/or Failsafe to use  `src/test/**kotlin**`, but I don't know how)
[x] Kotlin - Junit - Maven
[x] Kotlin - TestNg - Maven
[ ] Kotlin - Cli

Comments:
* Mainly minor textual edits; please accept if you agree.
* Question about running from the CLI
* The part about the jars really needs to be removed. (Could you mention where to find the jars if they're downloaded with Maven?
* A few edits for the Kotlin part; please accept.

**Suggestion:** Please consider splitting this PR and moving the CLI part to a separate PR, so we can merge the JUnit/TestNG parts (with Kotlin edits), which imho are a great addition!

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+
+class Stepdefs : En {
+		
+		Given("Step from {string} in {string} feature file") { scenario: String , file: String ->
```
```suggestion
		    init {
        Given("Step from {string} in {string} feature file") { scenario: String , file: String ->
            println("Thread ID - ${Thread.currentThread().id} - $scenario from $file feature file")
        }
    }
```

mlvandijk replied:
When using lambda's with Kotlin, the stepdefs need to be inside an `init {}` block.
We could consider adding examples both with annotations and lambdas, but that could be a separate additional PR; let's get this one merged asap. :)

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+import cucumber.api.java.BeforeStep;
+import cucumber.api.java.en.Given;
+
+public class Stepdefs {
```
```suggestion
public class StepDefs {
```

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+
+import cucumber.api.java8.En
+
+class Stepdefs : En {
```
```suggestion
class StepDefs : En {
```

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,380 @@
+- Compile the step definition class. Add the **path to the folder containing cucumber jars to the classpath** using the **-cp** option.
+
+```shell
+javac -cp .;<path to cucumber jar folder>/* ./parallel/Stepdefs.java
````
```suggestion
javac -cp .;<path to cucumber jar folder>/* ./parallel/StepDefs.java
```

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,380 @@
+- Compile the step definition class. Add **path to each of the downloaded cucumber jars to the classpath** using the **-cp** option.
+
+```shell
+kotlinc -cp .;<path to each cucumber jar> -jvm-target 1.8 ./parallel/Stepdefs.java
````
```suggestion
kotlinc -cp .;<path to each cucumber jar> -jvm-target 1.8 ./parallel/StepDefs.kt
```

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+import org.junit.runner.RunWith
+
+@RunWith(Cucumber::class)
+class RunCucumberTest {
```
```suggestion
class RunCucumberTest
```

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+
+@RunWith(Cucumber::class)
+class RunCucumberTest {
+}
```
```suggestion

```

mlvandijk replied:
These brackets are optional in Kotlin for empty class.

grasshopper7 replied:
Added to batch commit

mlvandijk commented on content/docs/guides/parallel-execution.md:
```diff
@@ -0,0 +1,380 @@
+- Add a cucumber **runner** using the `RunWith` annotation in the `parallel` package (same name as step definition package) in the `src/test/java` folder.
+{{% /block %}}
+
+{{% block "java" %}}
```
Afaik you don't need both the {{% block "java" %}} and the ```java around the code for it to only show up in the right language tab, but do try that out first..

mlvandijk reviewed (commented):
Works with Kotlin, but does need a few edits in the Kotlin code / sections.

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,380 @@
+	</configuration>
+</plugin>
+```
+
````
{{% block "kotlin" %}}
For Surefire to find your step definitions, make sure they are in `src/test/**java**`.
{{% /block %}}

mlvandijk replied:
Alternatively, it might be possible to configure Surefire to use `src/test/kotlin`, but I don't know how.

mlvandijk commented on content/docs/guides/parallel-execution.md:
````diff
@@ -0,0 +1,380 @@
+```
+
+- To execute using a Maven **Failsafe plugin include the below configuration** in the `build` section to the `POM`. Rename the runner class to `RunCucumberIT`.  You can find further details [here](/docs/community/not-cucumber/#maven-execution-plugin).
+
````
```suggestion
{{% block "kotlin" %}}
For Failsafe to find your step definitions, make sure they are in src/test/**java**.
{{% /block %}}
```

mlvandijk replied:
Alternatively, it might be possible to configure Failsafe to use `src/test/kotlin`, but I don't know how.

mlvandijk reviewed (commented):
Additional Kotlin edits.

grasshopper7 committed:
Apply suggestions from code review

Text and Kotlin edits from review.

Co-Authored-By: Marit van Dijk <mlvandijk@gmail.com>

mlvandijk committed:
Minor edits for Kotlin

mlvandijk reviewed (approved):
Thanks so much for all your work on this @grasshopper7 - really appreciate it!!

grasshopper7 commented:
@mlvandijk I found some issues in this. Just fixing them. Will create a new PR.

mlvandijk commented:
Sure, thanks!
